packages/nx/src/plugins/js/package-json/create-package-json.spec.ts

```diff
@@ -499,4 +503,214 @@ describe('createPackageJson', () => {
+        },
+      },
+      dependencies: {
+        app1: appDependencies,
+        lib1: libDependencies,
+      },
+    };
+
+    const rootPackageJson = () => ({
+      dependencies: {
+        '@nx/devkit': '~16.0.0',
+        nx: '> 14',
+        typescript: '^4.8.2',
+        tslib: '~2.4.0',
+      },
+    });
+
+    const projectPackageJson = () => ({
+      name: 'other-name',
+      version: '1.2.3',
+      dependencies: {
+        typescript: '^4.8.4',
+        random: '1.0.0',
+      },
+    });
+
+    const spies = [];
+
+    afterEach(() => {
+      while (spies.length > 0) {
+        spies.pop().mockRestore();
+      }
+    });
+
+    it('should use fixed versions when creating package json for apps', () => {
+      spies.push(
+        jest
+          .spyOn(fileutilsModule, 'readJsonFile')
+          .mockImplementation((path) => {
+            if (path === '/root/package.json') {
+              return rootPackageJson();
+            }
+          })
+      );
+
+      expect(createPackageJson('app1', graph)).toEqual({
+        dependencies: {
+          '@nx/devkit': '16.0.0',
+          nx: '16.0.0',
+          typescript: '4.9.5',
+        },
+        name: 'app1',
+        version: '0.0.1',
+      });
+    });
+
+    it('should override fixed versions with local ranges when creating package json for apps', () => {
+      spies.push(
+        jest.spyOn(fs, 'existsSync').mockImplementation((path) => {
+          if (path === '/root/apps/app1/package.json') {
+            return true;
+          }
+        })
+      );
+      spies.push(
+        jest
+          .spyOn(fileutilsModule, 'readJsonFile')
+          .mockImplementation((path) => {
+            if (path === '/root/package.json') {
+              return rootPackageJson();
+            }
+            if (path === '/root/apps/app1/package.json') {
+              return projectPackageJson();
+            }
+          })
+      );
+
+      expect(createPackageJson('app1', graph)).toEqual({
+        dependencies: {
+          '@nx/devkit': '16.0.0',
+          nx: '16.0.0',
+          random: '1.0.0',
+          typescript: '^4.8.4',
+        },
+        name: 'other-name',
+        version: '1.2.3',
+      });
+    });
+
+    it('should use range versions when creating package json for libs', () => {
+      spies.push(
+        jest
+          .spyOn(fileutilsModule, 'readJsonFile')
+          .mockImplementation((path) => {
+            if (path === '/root/package.json') {
+              return rootPackageJson();
+            }
+          })
+      );
+
+      expect(createPackageJson('lib1', graph)).toEqual({
+        dependencies: {
+          '@nx/devkit': '~16.0.0',
+          tslib: '~2.4.0',
+          typescript: '^4.8.2',
+        },
+        name: 'lib1',
+        version: '0.0.1',
+      });
+    });
+
+    it('should override range versions with local ranges when creating package json for libs', () => {
+      spies.push(
```

Comment:
Not a fan of having to mock these functions, but since it's used in existing tests let's go with it for now. Maybe something we can discuss later is testing strategies